docs/send-finality-vote.md

```diff
@@ -0,0 +1,82 @@
+# Finality Votes Submission Specification
+
+## Overview
+
+Finality providers submit votes to finalize blocks on the consumer chain.
+This document specifies the process of submitting finality votes.
+
+## Internal State
+
+The finality provider maintains two critical persistent states:
```

Comment:
i think it's simpler to just keep one internal state i.e. last processed height

last voted height can be easily retrieved from remote

Reply:
actually do you think we can get rid of both from DB and only keeps one in-memory state (i.e. `fp.nextHeightToProcess`)?

I feel it's doable

Reply:
Good point. I think we maybe can get rid of `lastVotedHeight` but why do you say it's easy to retrieve `lastVotedHeight` from the remote? It seems that we need to save this info in the finality provider on Babylon

> actually do you think we can get rid of both from DB and only keeps one in-memory state (i.e. fp.nextHeightToProcess)?

if we get rid of `fp.nextHeightToProcess`, we will waste some resources processing blocks that have already been processed, won't we?

Reply:
> but why do you say it's easy to retrieve lastVotedHeight from the remote? It seems that we need to save this info in the finality provider on Babylon

b/c if we store it in a local DB, there will be two source of truth. for the OP consumer FP, we can easily add a query in the CosmWasm contract for it e.g. `getFpLastVotedHeight`

Reply:
> if we get rid of fp.nextHeightToProcess, we will waste some resources processing blocks that have already been processed, won't we?

no I mean to get rid of the local DB states but keep `nextHeightToProcess`

how it works is: at startup, the FP will get everything it needs from the Babylon Chain and/or the CosmWasm contract to set `nextHeightToProcess`

but `nextHeightToProcess` is not a DB state, it's an in-memory field value of the FinalityProvider object

Reply:
> b/c if we store it in a local DB, there will be two source of truth. for the OP consumer FP, we can easily add a query in the CosmWasm contract for it e.g. getFpLastVotedHeight

Does CosmWasm contract maintain `lastVotedHeight` for query? I don't think Babylon stores such information. I think it's worth adding it. Wdyt @SebastianElvis?

Reply:
> how it works is: at startup, the FP will get everything it needs from the Babylon Chain and/or the CosmWasm contract to set nextHeightToProcess

If I understand it correctly, we only have `lastVotedHeight` to update the in-memory field `nextHeightToProcess`, right? If the fp does not have voting power for a lot of blocks, after processing them and restart, the the fp needs to process those block again

Reply:
These are specific for Consumer. Each consumer chain deploys its own CW contract storing the data

Reply:
For Babylon FP, you can probably add a similar query API directly

Reply:
> If I understand it correctly, we only have `lastVotedHeight` to update the in-memory field `nextHeightToProcess`, right?

maybe I misunderstood, I thought you are proposing to have two DB states:

- Last voted height
- Last processed height

my proposal is we don't need neither to update the in-memory field `nextHeightToProcess`. we always read from remote at startup time

>  If the fp does not have voting power for a lot of blocks, after processing them and restart, the the fp needs to process those block again

right this is an edge case which I think won't happen too frequently. you need to satisfy: 1) FP doesn't have power; 2) You restart the FP when it doesn't have power. This doesn't see a frequent use case.

I'd vote instead to keep the architecture simple than optimize for it.

Reply:
> For Babylon FP, you can probably add a similar query API directly

I don't think so as `LastVotedHeight` is not stored anywhere, but I think it's worth storing it.

> my proposal is we don't need neither to update the in-memory field nextHeightToProcess. we always read from remote at startup time

I meant we can only read `lastVotedHeight` from the remote to update `nextHeightToProcess`

> right this is an edge case which I think won't happen too frequently. you need to satisfy: 1) FP doesn't have power; 2) You restart the FP when it doesn't have power. This doesn't seem like a frequent use case.

It's not a frequent case but it's a valid case and persisting `lastProcessedHeight` can address it easily. And, the inconsistency of `lastProcessedHeight` does not hurt security. So I would rather keep it as it is now unless you think it's not secure to do so

Reply:
Makes sense. Will bring it to the team to make the decision. Appreciate the discussion!

Reply:
Yep, I don't think it's safe to assume that FP is connected to a trusted RPC as we can not control it. The bottom line is that fp should not sign twice for the same height.